refactor AdjacencyBitmap to AdjacencyMap without bitwise operations

Em3rgencyLT opened the pull request:
## Summary
This simplifies working with the AdjacencyMap for future additions of different adjacency types. It adds much needed diagonal adjacency support without needing to do hacky workarounds. It also improves readability in several key areas, at the cost of some verboseness. Performance should not be affected much, key changes made were from using bitwise comparisons to an array of structs.

* remove bitwise logic for determining neighbouring tile adjacency
* add diagonal direction support for AdjacencyMap
* fix door.cs not working correctly with tilemap editor
* minor style fixes and cleanup on some touched files

## Changes to Files
~Biggest change was converting all the AdjacencyBitmap logic where a byte stored the 8 directions connection information to AdjacencyMap where a Dictionary<Direction, bool> stores it. Everything else was just updating the dependencies that used AdjacencyBitmap before.~

Created AdjacencyData struct that contains 3 fields - generic type, specific type and exists. We will need to track what kind of type the adjacency is for some of the new connectors. The types themselves aren't implemented in this PR as it is out of scope.

Converted AdjacencyBitmap (now AdjacencyMap) logic to use an array of AdjacencyData structs instead of a byte. And updates all files that rely on AdjacencyBitmap.

## Technical Notes
~I chose a dictionary as it seemed the most appropriate structure for the adjacency data. I wanted an IEnumerable for easy filtering with Linq. Additionally directly accessing the values by using a Direction key is much more human readable (if somewhat less efficient) than a byte.~

~Regarding performance, I didn't perceive any change. Granted ministation is a small map. If this stops being scalable for networking when loading a bigger station, the main way to improve it would be to add areas of concern for clients. IE don't load the whole map, but only some surrounding chunks somewhat outside their view distance.~

Got rid of the dictionary implementation as it was too overengineered and caused a needless hit on performance. Now using an array, initializing the data for 1 million tiles only adds 1 second of extra processing time. A level of magnitude increase over the bitwise implementation which only took some few hundred ms, but we need to move away from byte either way and anything we use will be less efficient.

Test results comparison:
1 million tiles:
```
time to set bitwise connection 00:00:00.1049098
time in bitwise_test 00:00:00.1345397
time to set array connection 00:00:00.3721661
time in array_test 00:00:01.0938453
```
7 million tiles:
```
time to set bitwise connection 00:00:00.7761320
time in bitwise_test 00:00:01.5036310
time to set array connection 00:00:02.5088772
time in array_test 00:00:07.3375517
```

## Fixes
Closes #865

Em3rgencyLT commented:
Test coverage commit: https://github.com/Em3rgencyLT/SS3D/commit/d28d0a5be399622d3c45df156c18b84ad853f726

Will open PR after this merges. Don't want to bloat this PR with it.

stilnat commented:
This looks like great work, however, since it affects performances negatively, we should quantify this impact before merging.

Em3rgencyLT commented:
> This looks like great work, however, since it affects performances negatively, we should quantify this impact before merging. @Em3rgencyLT I'll try something but if you feel like providing your own tests that would be great !

What would be a satisfactory test? Not really sure how to quantify this sort of thing. If you have an idea i would be grateful if you did it. Additionally, If changes are required don't hesitate to let me know.

Em3rgencyLT commented:
As an extra piece of information, i dont think we can keep byte to track adjacency either way. Since for future features, keeping track of each ajacencies type and subtype is also required. My dictionary type will probably change to <Direction, Object>.

stilnat commented:
So I ran a few tests on your code, to compare current implementation and yours. My current results are that I can set 7 millions directions in 0.46 seconds with your code and 0.19 with the old one. It's not that bad considering that realistically the number of actual connections updated will be much less at any given time.
However when including initialisation time, the code performs very poorly. 0.23 seconds for 1 million tiles in the old code versus more than 5 seconds for your implementation. 

here's the code : 

`using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace test_bitwise_operator_vs_dictionnary
{

    public enum Direction
    {
        North = 0,
        NorthEast = 1,
        East = 2,
        SouthEast = 3,
        South = 4,
        SouthWest = 5,
        West = 6,
        NorthWest = 7,
    }
    
    class Program
    {

        static TimeSpan TimeAction(Action blockingAction)
        {
            System.Diagnostics.Stopwatch stopWatch = System.Diagnostics.Stopwatch.StartNew();
            blockingAction();
            stopWatch.Stop();
            return stopWatch.Elapsed;
        }


        static void Main(string[] args)
        {
            int nb = 1000000;

            var elapsed = TimeAction(() =>
            {
                bitwise_test(nb);
            });

            Console.WriteLine("time in bitwise_test " + elapsed);

            var elapsed2 = TimeAction(() =>
            {
                dictionnary_test(nb);     
            });

            Console.WriteLine("time in bitwise_test " + elapsed2);

        }

        static void dictionnary_test(int nb)
        {
            AdjacencyMap[] array1 = new AdjacencyMap[nb];
            for(int i=0; i< nb; i++)
            {
                array1[i] = new AdjacencyMap();
            }


            var elapsed = TimeAction(() =>
            {
                foreach (AdjacencyMap map in array1)
                {
                    map.SetConnection(Direction.NorthEast, true);
                    map.SetConnection(Direction.South, true);
                    map.SetConnection(Direction.West, true);
                    map.SetConnection(Direction.SouthWest, true);
                    map.SetConnection(Direction.West, false);
                    map.SetConnection(Direction.South, false);
                    map.SetConnection(Direction.NorthEast, false);
                }
            });

            Console.WriteLine("time to set connection " + elapsed);
        }

        static void bitwise_test(int nb)
        {

            AdjacencyBitmap[] array1 = new AdjacencyBitmap[nb];
            for (int i = 0; i < nb; i++)
            {
                array1[i] = new AdjacencyBitmap();
            }


            var elapsed = TimeAction(() =>
            {
                foreach (AdjacencyBitmap map in array1)
            {
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.NorthEast, true);
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.South, true);
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.West, true);
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.SouthWest, true);
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.West, false);
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.South, false);
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.NorthEast, false);
            }
            });

            Console.WriteLine("time to set connection " + elapsed);
        }

        public class AdjacencyMap
        {
            //Stores a map of which of the 8 surrounding tiles have a connection
            private Dictionary<Direction, bool> connections;

            public AdjacencyMap()
            {
                connections = new Dictionary<Direction, bool>();
                Enum.GetValues(typeof(Direction)).Cast<Direction>().ToList().ForEach(direction =>
                {
                    connections.Add(direction, false);
                });
            }

            public Dictionary<Direction, bool> Connections => connections;

            public bool HasConnection(Direction direction)
            {
                return connections[direction];
            }

            public int GetCardinalConnectionCount()
            {
                return GetAdjacensies().Count;
            }

            public int GetDiagonalConnectionCount()
            {
                return GetAdjacensies().Count;
            }

            /**
             * Gets the direction of the only cardinal/diagonal connection.
             * Assumes there is exactly one cardinal/diagonal connection.
             */
            public Direction GetSingleConnection(bool cardinal = true)
            {
                List<Direction> foundConnections = GetAdjacensies();
                return foundConnections[0];
            }

            /**
             * Get the direction of the only cardinal/diagonal non connection.
             * Assumes there is exactly one cardinal/diagonal non connection
             */
            public Direction GetSingleNonConnection(bool cardinal = true)
            {
                List<Direction> foundConnections = GetAdjacensies();
                List<Direction> directions = CardinalDirections() ;
                List<Direction> missingConnections =
                    directions.Where(direction => !foundConnections.Contains(direction)).ToList();

                return missingConnections[0];
            }

            /**
             * Updates the map, and returns whether there was a change
             */
            public bool SetConnection(Direction direction, bool value)
            {
                bool changed = connections[direction] != value;
                if (changed)
                {
                    connections[direction] = value;
                }

                return changed;
            }

            public bool SetConnections(Dictionary<Direction, bool> newConnections)
            {
                bool changed = !connections.Equals(newConnections);
                connections = newConnections;

                return changed;
            }

            public static List<Direction> CardinalDirections()
            {
                return new List<Direction> { Direction.North, Direction.East, Direction.South, Direction.West };
            }

            private List<Direction> GetAdjacensies()
            {
                List<Direction> directions = CardinalDirections();
                return connections
                    .Where(connection => connection.Value)
                    .Select(connection => connection.Key)
                    .Where(direction => directions.Contains(direction)).ToList();
            }
        }

        public class AdjacencyBitmap
        {
            public struct CardinalInfo
            {
                public int north;
                public int east;
                public int south;
                public int west;
                public int numConnections;

                public CardinalInfo(byte bitmap)
                {
                    north = Adjacent(bitmap, Direction.North);
                    east = Adjacent(bitmap, Direction.East);
                    south = Adjacent(bitmap, Direction.South);
                    west = Adjacent(bitmap, Direction.West);

                    numConnections = north + east + south + west;
                }

                /**
                 * Gets the direction of the only positive connection.
                 * Assumes there is exactly one positive connection
                 */
                public Direction GetOnlyPositive()
                {
                    return (Direction)(
                        north * (int)Direction.North
                        + east * (int)Direction.East
                        + south * (int)Direction.South
                        + west * (int)Direction.West
                    );
                }
                /**
                 * Get the direction of the only non-1 connection.
                 * Assumes there is exactly one non-1 connection
                 */
                public Direction GetOnlyNegative()
                {
                    return (Direction)(
                        (1 - north) * (int)Direction.North
                        + (1 - east) * (int)Direction.East
                        + (1 - south) * (int)Direction.South
                        + (1 - west) * (int)Direction.West
                    );
                }

                /**
                 * Assuming the cardinals have exactly two adjacent connections,
                 * gets the angle between them (will therefore be a diagonal angle).
                 */
                public Direction GetCornerDirection()
                {
                    return south > 0 ? east > 0 ? Direction.SouthEast : Direction.SouthWest : west > 0 ? Direction.NorthWest : Direction.NorthEast;
                }
            }

            /**
         * Returns a bitmap which modifies the input bitmap so that
         * the bit corresponding to the given direction is set to the given value.
         */
            public static byte SetDirection(byte bitmap, Direction direction, bool value)
            {
                // Set the direction bit to isConnected (1 or 0)
                byte output = bitmap;
                output &= (byte)~(1 << (int)direction);
                output |= (byte)((value ? 1 : 0) << (int)direction);

                return output;
            }

            public byte Connections { get; set; } = 0;

            /**
             * Updates the given bitmap using SetDirection, and returns whether the
             * bitmap changed as a result.
             * 
             * onlyCheckCardinals - If true, then only considers the bitmap has changed if N, E, S, or W has changed.
             */
            public bool UpdateDirection(Direction direction, bool value, bool onlyCheckCardinals = false)
            {
                byte output = SetDirection(Connections, direction, value);

                byte comparisonMap = onlyCheckCardinals ? (byte)0b01010101 : (byte)0b11111111;
                bool changed = (output & comparisonMap) != (Connections & comparisonMap);
                Connections = output;
                return changed;
            }

            /**
             * Gets (as 0 or 1) whether each cardinal direction is connected (N, E, S, W),
             * and the last value is the total number of cardinal connections
             */
            public CardinalInfo GetCardinalInfo()
            {
                return new CardinalInfo(Connections);
            }

            /**
             * Returns 0 if no adjacency, or 1 if there is.
             */
            public int Adjacent(Direction direction)
            {
                return Adjacent(Connections, direction);
            }

            public static int Adjacent(byte bitmap, Direction direction)
            {
                return (bitmap >> (int)direction) & 0x1;
            }
        }
    }
}
`

Em3rgencyLT commented:
Thank you for taking the time to do this @stilnat. This seems thorough and lifts the burden off me. What do you mean by 'including initialization'? Is that the time taken for memory allocation for the dictionaries or what?

Either way, a million tiles is a 1000x1000 station. Which is... Possible. The decision to be made then is - is 5 seconds+ an acceptable load time for such a station. Assuming we won't be chunking it and every player needs to load the entire thing.

Additionally, what is the alternative? Like I mentioned above, we need to begin tracking adjacency type and subtype for each connection. Is that something that we want to encode into bytes as well?

Em3rgencyLT commented:
Updated to use Array instead of Dictionary. Updated PR description. Re-ran the tests. Test code:
```
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace adjacencyMap
{
    public enum Direction
    {
        North = 0,
        NorthEast = 1,
        East = 2,
        SouthEast = 3,
        South = 4,
        SouthWest = 5,
        West = 6,
        NorthWest = 7,
    }
    
    public struct AdjacencyData
    {
        public AdjacencyData(string genericType, string specificType, bool exists)
        {
            GenericType = genericType;
            SpecificType = specificType;
            Exists = exists;
        }

        public string GenericType;
        public string SpecificType;
        public bool Exists;

        public override bool Equals(object other)
        {
            return other is AdjacencyData otherData && GenericType == otherData.GenericType && SpecificType == otherData.SpecificType &&
                   Exists == otherData.Exists;
        }
    }
    
    internal class Program
    {
        static TimeSpan TimeAction(Action blockingAction)
        {
            System.Diagnostics.Stopwatch stopWatch = System.Diagnostics.Stopwatch.StartNew();
            blockingAction();
            stopWatch.Stop();
            return stopWatch.Elapsed;
        }
        
        public static void Main(string[] args)
        {
            int nb = 1000000;

            var elapsed = TimeAction(() =>
            {
                bitwise_test(nb);
            });

            Console.WriteLine("time in bitwise_test " + elapsed);

            var elapsed2 = TimeAction(() =>
            {
                array_test(nb);     
            });

            Console.WriteLine("time in array_test " + elapsed2);
        }
        
         static void array_test(int nb)
        {
            AdjacencyMap[] array1 = new AdjacencyMap[nb];
            for(int i=0; i< nb; i++)
            {
                array1[i] = new AdjacencyMap();
            }
            
            var elapsed = TimeAction(() =>
            {
                foreach (AdjacencyMap map in array1)
                {
                    map.SetConnection(Direction.NorthEast, new AdjacencyData("", "", true));
                    map.SetConnection(Direction.South, new AdjacencyData("", "", true));
                    map.SetConnection(Direction.West, new AdjacencyData("", "", true));
                    map.SetConnection(Direction.SouthWest, new AdjacencyData("", "", true));
                    map.SetConnection(Direction.West, new AdjacencyData("", "", false));
                    map.SetConnection(Direction.South, new AdjacencyData("", "", false));
                    map.SetConnection(Direction.NorthEast, new AdjacencyData("", "", false));
                }
            });

            Console.WriteLine("time to set array connection " + elapsed);
        }

        static void bitwise_test(int nb)
        {

            AdjacencyBitmap[] array1 = new AdjacencyBitmap[nb];
            for (int i = 0; i < nb; i++)
            {
                array1[i] = new AdjacencyBitmap();
            }


            var elapsed = TimeAction(() =>
            {
                foreach (AdjacencyBitmap map in array1)
            {
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.NorthEast, true);
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.South, true);
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.West, true);
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.SouthWest, true);
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.West, false);
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.South, false);
                map.Connections = AdjacencyBitmap.SetDirection(map.Connections, Direction.NorthEast, false);
            }
            });

            Console.WriteLine("time to set bitwise connection " + elapsed);
        }
        
        public class AdjacencyMap
    {
        //Stores an array of which of the 8 surrounding tiles have a connection. Order assumed to match Direction enum values.
        private AdjacencyData[] connections;

        public AdjacencyMap()
        {
            connections = new [] {
                new AdjacencyData ("", "", false),
                new AdjacencyData ("", "", false),
                new AdjacencyData ("", "", false),
                new AdjacencyData ("", "", false),
                new AdjacencyData ("", "", false),
                new AdjacencyData ("", "", false),
                new AdjacencyData ("", "", false),
                new AdjacencyData ("", "", false),
            };
        }

        /**
         * Updates the map, and returns whether there was a change
         */
        public bool SetConnection(Direction direction, AdjacencyData data)
        {
            bool changed = !data.Equals(connections[(int) direction]);
            if (changed)
            {
                connections[(int) direction] = data;
            }

            return changed;
        }

        public override string ToString()
        {
            return $"{connections}";
        }
    }
        
         public class AdjacencyBitmap
        {
            public struct CardinalInfo
            {
                public int north;
                public int east;
                public int south;
                public int west;
                public int numConnections;

                public CardinalInfo(byte bitmap)
                {
                    north = Adjacent(bitmap, Direction.North);
                    east = Adjacent(bitmap, Direction.East);
                    south = Adjacent(bitmap, Direction.South);
                    west = Adjacent(bitmap, Direction.West);

                    numConnections = north + east + south + west;
                }
            }

            /**
         * Returns a bitmap which modifies the input bitmap so that
         * the bit corresponding to the given direction is set to the given value.
         */
            public static byte SetDirection(byte bitmap, Direction direction, bool value)
            {
                // Set the direction bit to isConnected (1 or 0)
                byte output = bitmap;
                output &= (byte)~(1 << (int)direction);
                output |= (byte)((value ? 1 : 0) << (int)direction);

                return output;
            }

            public byte Connections { get; set; } = 0;


            public static int Adjacent(byte bitmap, Direction direction)
            {
                return (bitmap >> (int)direction) & 0x1;
            }
        }
    }
}```

stilnat commented on Assets/Engine/Tile/TileRework/Connections/AdjacencyMap.cs:
```diff
@@ -0,0 +1,146 @@
+            connections = newConnections;
+        }
+
+        private List<Direction> GetAdjacensies(bool cardinal)
```
could you rename it in GetAdjacencies ?

stilnat reviewed (changes requested):
Great work ! On my side I'll be happy to accept the changes, there's just a small english typo to correct with "GetAdjacensies" :) .

Em3rgencyLT commented:
Done

stilnat reviewed (approved):
Great work ! This make a solid base to work on future connectors, everything checks out !
